### PR TITLE
Cached metric value

### PR DESCRIPTION
**Pull request by chuenlok**
This bases on subcard fix branch.

### CODE REVIEW

File: mod/wikirate/set/type_plus_right/metric/all_values.rb [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+  end
+  @query
+end
+
+def get_cached_result
+  @cached_metric_values ||= begin
+    cached_json = fetch(trait: :cached_count, new: {}).format.render_core
+    JSON.parse(cached_json)
+  end
+end
```
Comment (ethn) [inline mod/wikirate/set/type_plus_right/metric/all_values.rb:51]: Wondering if we shouldn't generalize this the same way we generalize CachedCount.  (Perhaps call it CachedSearchResult)?

File: mod/wikirate/set/type_plus_right/metric/all_values.rb [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+def count params={}
+  get_cached_result.size
+end
```
Comment (ethn) [inline mod/wikirate/set/type_plus_right/metric/all_values.rb]: if we're doing this, why are we including CachedCount?

File: mod/wikirate/set/type_plus_right/metric/all_values.rb [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+  view :card_list_items do |args|
+    results =
+      search_results.map do |row|
+        c = Card["#{card.cardname.left}+#{row[0]}"]
+        render :card_list_item, args.clone.merge(item_card: c)
+      end.join "\n"
+  end
+
+  view :card_list do |args|
+    paging = _optional_render :paging, args.merge({cached_query: true})
```
Comment (ethn) [inline mod/wikirate/set/type_plus_right/metric/all_values.rb]: what's this "cached_query" arg about?

File: mod/wikirate/set/type_plus_right/metric/all_values.rb (hunk 1 above)
Comment (ethn) [inline mod/wikirate/set/type_plus_right/metric/all_values.rb:51]: wait, the cached_count is storing json results?

File: mod/wikirate/set/type_plus_right/metric/all_values.rb [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+        #{paging}
+        #{header}
+        <div class="search-result-list">
+          #{results}
+        </div>
+        #{paging if search_results.length > 10}
+      }
+    end
+  end
+end
```
Comment (ethn) [inline mod/wikirate/set/type_plus_right/metric/all_values.rb:206]: there is a whole lot of duplicate functionality here and in search_type.rb.  Must there be?

Comment (chuenlok) [inline mod/wikirate/set/type_plus_right/metric/all_values.rb:206]: I know it is duplicated but I need to insert `#{header}# between`#{paging}` and the results. Any better way to do that?

File: mod/wikirate/set/type_plus_right/metric/all_values.rb (hunk 1 above)
Comment (chuenlok) [inline mod/wikirate/set/type_plus_right/metric/all_values.rb:51]: Yes. Philipp said he will change the name from `cached_count` to something appropriate later.

File: mod/wikirate/set/type_plus_right/metric/all_values.rb [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+  item_cards(default_query: true).each do |card|
+    company_name = card.left.left.cardname.right
+    year = card.cardname.parts[-2]
+    value = card.content
+    unless result.has_key?company_name
+      result[company_name] = []
+    end
+    result[company_name].push({ year: year, value: value})
+  end
+  result.to_json
```
Comment (ethn) [inline mod/wikirate/set/type_plus_right/metric/all_values.rb:242]: It's clever to try to reuse this code, but we can't do this without serious renaming and/or refactoring.  We can't have a method called "calculate_count" return json with a bunch of values.  it's just too hacky.

File: mod/wikirate/set/type_plus_right/metric/all_values.rb (hunk 4 above)
Comment (ethn) [inline mod/wikirate/set/type_plus_right/metric/all_values.rb:206]: the header is the sorting column headers, right?

Comment (chuenlok) [inline mod/wikirate/set/type_plus_right/metric/all_values.rb:206]: yes. Should I update wagn to decompose the card_list view?

File: mod/wikirate/set/type_plus_right/metric/all_values.rb (hunk 5 above)
Comment (chuenlok) [inline mod/wikirate/set/type_plus_right/metric/all_values.rb:242]: I will try to the caching mod tmr

**Comment by chuenlok** [comment]
There are still problems on the paging. I will fix it soon

File: mod/wikirate/set/type_plus_right/metric/all_values.rb (hunk 3 above)
Comment (chuenlok) [inline mod/wikirate/set/type_plus_right/metric/all_values.rb]: it is removed now.

**Pull request merged**